### PR TITLE
[ozone/mus] Add support to create a new window by dragging a tab.

### PR DESCRIPTION
**Pull request by msisov**
[ozone/mus] Add inital support for a tab drag window.
    
tl;dr
This patch adds an initial support to start a move loop and create
a new window by dragging a tab. To start with, this commit
extends WindowTree::PerformWindowMove, which calls
WindowServer::StartMoveLoop -> PlatformWindow::RunMoveLoop.
  
X11 Ozone implementation of PlatformWindow uses
WindowMoveLoopClient, which instantiates the move loop and
WholeScreenMoveLoop, which creates an invisible window
and intercepts all the events from it. Then the system screen
location is taken and sent to WindowMoveLoopClient, which
updates actual bounds of X11WindowBase.
    
The way it works is precisly the same as in stock X11, but
further work to share the code is needed, because there are
some difference between stock x11 and ozone x11 event
handlings, bounds set and etc.

Known issues:
1) Dragging a tab out of a window and dragging it back results in broken capture. User has to
click on the non-client area (the topmost) to restore the capture.

2) Dragging a tab back and forth to detach and attach the tab, might result in corrupted widget,
which doesn't show anything, but some garbage.

3) Attaching a window into a tab strip might result to segfault for hit test aggregator.
Which is actually a race between switch hit test regions and destroyed ozone window and its components.

And full description:
The point where run loop starts in mus is DesktopWindowTreeHostMus::RunMoveLoop.
It calls to WindowTreeClient::PerformWindowMove, which asks WindowTree to do actual move loop.
Once the request goes through the access policy, WindowServer::StartMoveLoop is called.
Then ws::PlatformDisplay is taken and native PlatformWindow is asked to perform a move loop.

A move loop consists of several objects - X11WindowOzone, who instantiates the move loop on ozone side, WindowMoveLoopClient, who manages WholeScreenMoveLoop and receives mouse coordinates, and the mentioned WholeScreenMoveLoop, who creates an invisible window, sets an explicit mouse grab on it and starts actual RunLoop intercepting all the events, modifying their locations to system coordinates locations and sending them both to the next dispatchers (which get the events, propagate to aura and modify the actual mouse locations both in mus and aura side, which is extremely important for TabDragController to work, because it always checks where the pointer is,once the pointer is on top of the location with a tab strip, it can attach a window inside a tab strip and make a tab out of that) and WindowMoveLoopClient, which adds a drag offset to the coordinates it has received and updates the location of the X11WindowOzone.

Once WholeScreenMoveLoop receives a mouse release event or keyboard escape event, it stops the move loop and WholeScreenMoveLoop::RunMoveLoop returns the success of the operation to WindowTree. In turn, WindowTree calls OnChangeCompleted(result).
    
Issue #264

### CODE REVIEW

**Review by tonikitoo — APPROVED** [review]

File: services/ui/ws/window_tree.cc [hunk 1]
```diff
@@ -2465,6 +2466,14 @@ void WindowTree::PerformWindowMove(uint32_t change_id,
     return;
   }
 
+  if (window_server_->IsInExternalWindowMode()) {
+    // In external window mode, we just return early, which means an access to
+    // perform the window move is allowed.
```
Comment (tonikitoo) [inline services/ui/ws/window_tree.cc]: nit: the part "which means an access to perform the window mode" isnt too clear in my mind. maybe replace "access" with something else?

File: services/ui/ws/window_tree.cc [hunk 2]
```diff
@@ -2465,6 +2466,14 @@ void WindowTree::PerformWindowMove(uint32_t change_id,
     return;
   }
 
+  if (window_server_->IsInExternalWindowMode()) {
```
Comment (tonikitoo) [inline services/ui/ws/window_tree.cc]: maybe you could move this if above line 2444?

if (success && window_server_->IsInExternalWindowMode()) {
  // comment
  // todo
  OnChangeComplete(change_id, true);
  return;
}

Comment (msisov) [reply services/ui/ws/window_tree.cc]: Well, the if (success...) is redundant here. I left it by accident, but if I move it above 2444, then checks on lines 2454 and 2461 are going to be missed, which is not a right way to do.

**Review by tonikitoo — APPROVED** [review]
lgtm

File: services/ui/public/interfaces/window_tree.mojom [hunk 3]
```diff
@@ -320,12 +320,14 @@ interface WindowTree {
...
+  // Tells the (host) window manager to start moving the window. OnChangeCompleted is
   // called on whether the move was canceled. Because there's a delay between
   // when a client sends this message and when the window manager starts acting
-  // on it, pass the cursor location at the start of the move.
+  // on it, pass the cursor location at the start of the move. In external
+  // window move, |drag_offset| is passed to ozone backend for proper handling
+  // of the move loop.
   PerformWindowMove(uint32 change_id, uint32 window_id, MoveLoopSource source,
-                    gfx.mojom.Point cursor);
+                    gfx.mojom.Point cursor, gfx.mojom.Vector2d drag_offset);
```
Comment (tonikitoo) [inline services/ui/public/interfaces/window_tree.mojom:330]: question: I take it that chromeos still works ok with this change?

File: services/ui/ws/platform_display.cc [hunk 4]
```diff
@@ -34,5 +34,11 @@ std::unique_ptr<PlatformDisplay> PlatformDisplay::Create(
 #endif
 }
 
+bool PlatformDisplay::RunMoveLoop(const gfx::Vector2d& drag_offset){
```
Comment (tonikitoo) [inline services/ui/ws/platform_display.cc]: nit: space before "{"

File: services/ui/ws/window_server.cc [hunk 5]
```diff
@@ -507,18 +508,44 @@ void WindowServer::SetPaintCallback(
...
-void WindowServer::StartMoveLoop(uint32_t change_id,
+bool WindowServer::StartMoveLoop(uint32_t change_id,
                                  ServerWindow* window,
                                  WindowTree* initiator,
-                                 const gfx::Rect& revert_bounds) {
+                                 const gfx::Rect& revert_bounds,
+                                 const gfx::Vector2d& drag_offset) {
   current_move_loop_.reset(
       new CurrentMoveLoopState{change_id, window, initiator, revert_bounds});
+
```
Comment (tonikitoo) [inline services/ui/ws/window_server.cc:518]: Ideally, WindowServer should not know about ws::PlatformDIsplay, and route things through the ws::Display.

File: services/ui/ws/window_tree.cc [hunk 6]
```diff
@@ -2442,13 +2442,13 @@ void WindowTree::CancelDragDrop(Id window_id) {
   wms->CancelDragDrop();
 }
 
-void WindowTree::PerformWindowMove(uint32_t change_id,
-                                   Id window_id,
-                                   ui::mojom::MoveLoopSource source,
-                                   const gfx::Point& cursor) {
+void WindowTree::PerformWindowMove(uint32_t change_id, Id window_id,
```
Comment (tonikitoo) [inline services/ui/ws/window_tree.cc:2445]: suggestion: maybe it makes sense to add an early if-block for external window mode?

for example

in like 2450:

if (success && window_server_->IsInExternalWindowMode()) {
  // do our thing, and return
}

// and leave the rest of the method untouched?

File: services/ui/ws/window_tree_unittest.cc [hunk 7]
```diff
@@ -1325,7 +1325,7 @@ TEST_F(WindowTreeTest, MoveLoopAckOKByWM) {
       ->PerformWindowMove(
           change_id,
           child_tree->ClientWindowIdToTransportId(embed_window_id2_in_child),
-          mojom::MoveLoopSource::MOUSE, gfx::Point(0, 0));
+          mojom::MoveLoopSource::MOUSE, gfx::Point(0, 0), gfx::Vector2d());
```
Comment (tonikitoo) [inline services/ui/ws/window_tree_unittest.cc:1328]: suggestion: maybe instead of adding the mechanical default last parameter value here, we could declare the method with a default value?

then the call sites would not need change. does it make sense?

File: services/ui/public/interfaces/window_tree.mojom (hunk 3 above)
Comment (msisov) [reply services/ui/public/interfaces/window_tree.mojom:330]: right.

File: services/ui/ws/platform_display.cc (hunk 4 above)
Comment (msisov) [reply services/ui/ws/platform_display.cc]: done

File: services/ui/ws/window_server.cc (hunk 5 above)
Comment (msisov) [reply services/ui/ws/window_server.cc:518]: ok, let me then modify other callers and do this one in the follow ups!

File: services/ui/ws/window_tree.cc (hunk 6 above)
Comment (msisov) [reply services/ui/ws/window_tree.cc:2445]: as per discussion, we modify change ids and rely on the checks perform before the call to window_server_->RunMoveLoop. so, leave the code as it is.

File: services/ui/ws/window_tree_unittest.cc (hunk 7 above)
Comment (msisov) [reply services/ui/ws/window_tree_unittest.cc:1328]: leaving as it is because defaulting requires too much to do still.

File: services/ui/public/interfaces/window_tree.mojom [hunk 8]
```diff
@@ -320,12 +320,14 @@ interface WindowTree {
   // where the location is (x << 16) | y.
   GetCursorLocationMemory() => (handle<shared_buffer> cursor_buffer);
 
-  // Tells the window manager to start moving the window. OnChangeCompleted is
+  // Tells the (host) window manager to start moving the window. OnChangeCompleted is
   // called on whether the move was canceled. Because there's a delay between
   // when a client sends this message and when the window manager starts acting
-  // on it, pass the cursor location at the start of the move.
+  // on it, pass the cursor location at the start of the move. In external
+  // window move, |drag_offset| is passed to ozone backend for proper handling
```
Comment (rjkroege) [inline services/ui/public/interfaces/window_tree.mojom:327]: So, what should `drag_offset` be in CrOS? This is a suggestion to offer to the window manager?

File: services/ui/ws/platform_display.cc [hunk 9]
```diff
@@ -34,5 +34,11 @@ std::unique_ptr<PlatformDisplay> PlatformDisplay::Create(
 #endif
 }
 
+bool PlatformDisplay::RunMoveLoop(const gfx::Vector2d& drag_offset) {
```
Comment (rjkroege) [inline services/ui/ws/platform_display.cc:37]: How do the events flow? In particular, why can't `mus` short-circuit the event flow and listen to the incoming event stream to position the window? Is this is a dumb idea?

File: ui/platform_window/whole_screen_move_loop.h [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+
+
+
+namespace ui {
+class MouseEvent;
+class ScopedEventDispatcher;
+class XScopedEventSelector;
+class PlatformWindowDelegate;
+
+// Runs a nested run loop and grabs the mouse. This is used to implement
```
Comment (rjkroege) [inline ui/platform_window/whole_screen_move_loop.h:31]: It's not clear to me how this wires into the larger world. Can you explain?

**Review by rjkroege — COMMENTED** [review]
Some questions. Conceivably foolish. Hopefully you'll educate me.

File: services/ui/public/interfaces/window_tree.mojom (hunk 8 above)
Comment (msisov) [reply services/ui/public/interfaces/window_tree.mojom:327]: the drag_offset is the offset for the mouse pointer position, which is used to calculate where a being dragged window should be placed. It is passed to WindowMoveLoopClient, which stores the value and uses it on new mouse move/drag events (on WindowMoveLoopClient::OnMouseMovement) . You can check it here https://github.com/Igalia/chromium/pull/277/commits/4101041e5bf5d5f5052b288a04f53056f308ec51#diff-73743bddabef9dc55d92578838743a5aR31 

ChromeOS does things differently - in case of mus, ToplevelWindowEventHandler::RunMoveLoop is called instead of DesktopWindowTreeHostMus::RunMoveLoop. It doesn't use the drag offset, but instead calls to GetLastTouchPointForTarget or GetLastMouseLocationInRoot to get drag location.

ChromeOs mash does go through DesktopWindowTreeHostMus::RunMoveLoop, but in calls to window_manager with cursor location that translates the cursor to a root window and calculates bounds on drag using that offset.

File: services/ui/ws/platform_display.cc (hunk 9 above)
Comment (msisov) [reply services/ui/ws/platform_display.cc:37]: Well, I've tried that approach, but the movement of the window wasn't smooth enough. It was jumping back and forth while dragging.

File: ui/platform_window/whole_screen_move_loop.h (hunk 10 above)
Comment (msisov) [reply ui/platform_window/whole_screen_move_loop.h:31]: Ok, the WholeScreenMoveLoop does the same stuff as X11WholeScreenMoveLoop except some parts. For example, the events are handled differently, and it's initialized a bit differently. 

That object does the following - it creates an invisible xwindow and maps all the events to it by explicit grab. Wherever a mouse pointer is moved, all the events go to that object, which modifies events to have a root location as their location (this is done for a proper mouse locations update in Env, because TabDragController uses the last mouse pointer events location to know when it can attach the being dragged window to a tab strip). Then the location of the pointer is passed to WindowMoveLoopClient, which updates X11WindowOzone bounds.

Comment (msisov) [reply ui/platform_window/whole_screen_move_loop.h:31]: Once we get release event, run loop is stopped and WindowTree::PerformWindowMove() calls OnChangeCompleted with a result of the move loop.
